protocols/request-response/src/codec.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+pub use libp2p_core::ProtocolName;
+
+use futures::{prelude::*, future::BoxFuture};
+use std::io;
+
+/// A `RequestResponseCodec` defines the request and response types
+/// for a [`RequestResponse`](crate::RequestResponse) protocol or
+/// protocol family and how they are encoded / decoded on an I/O stream.
+pub trait RequestResponseCodec {
+    /// The type of protocol(s) or protocol versions being negotiated.
+    type Protocol: ProtocolName + Send + Sync + Clone;
+    /// The type of inbound and outbound requests.
+    type Request: Send + Clone;
+    /// The type of inbound and outbound responses.
+    type Response: Send;
+
+    /// Reads a request from the given I/O stream according to the
+    /// negotiated protocol.
+    fn read_request<'a, T>(&mut self, protocol: &Self::Protocol, io: &'a mut T)
+        -> BoxFuture<'a, Result<Self::Request, io::Error>>
+    where
+        T: AsyncRead + Unpin + Send;
+
+    /// Reads a response from the given I/O stream according to the
+    /// negotiated protocol.
+    fn read_response<'a, T>(&mut self, protocol: &Self::Protocol, io: &'a mut T)
+        -> BoxFuture<'a, Result<Self::Response, io::Error>>
```

Comment:
I'm personally not a fan of `async-trait`. It hides things from you for a very minimal gain.

Reply:
Thanks for the suggestion. I wasn't sure how controversial it would be and was even curious if someone would suggest it. While I have no strong preference and it is a relatively small convenience, I do see the appeal in getting syntactic uniformity for async (i.e. future-returning) methods, whether in traits or not. In terms of hiding things I personally don't think it makes the situation worse because the code expands in a very similar way to that of the existing built-in `async fn`s for non-trait methods (subject to the inherent limitations, of course), which one needs to understand in any case. So I have `async-trait` a try: https://github.com/libp2p/rust-libp2p/pull/1596/commits/9c6e6b4e88de7f7a7369b3f3834ab5f5233bc398. If there are strong objections, please let me know.